File: db/doc/ruby/money.rb
require "date.rb"

def pickNums(fromLst, count)
    result = []
    while result.size < count do
        num = fromLst[rand fromLst.size]
        result.push num if !result.include? num
    end
    result
end

def save2File(fp, dataLst)
    if fp
        resStr = (dataLst.join " ") + "\n"
        fp.print resStr 
        print resStr
    end
end

#
#            产生双色球号码
#   count:   注数
#   noFst:   不要的红区号列表
#   noSnd:   不要的蓝区号列表
#
def winSSQ(count, noRed, noBlue)
    #yesRed = (1..33).to_a.reject {|n| noRed.include? n}
    yesRed = good_red.reject {|n| noRed.include? n}
    yesBlue = (1..16).to_a.reject {|n| noBlue.include? n}
    srand
    begin
        f = open('ssqNum.txt', 'w')
        count.times {
            result = pickNums(yesRed,6) + pickNums(yesBlue,1)
            save2File(f, result)
        }
    ensure
        f.close if f
    end
end

def good_red
    sumRedNum = {}; (1..33).to_a.each {|n| sumRedNum[n] = 0}
    File.foreach("ssqHitNum.txt") { |line|
        num = (line.split " ")[1..7].map {|x| x.to_i}
        numRed = num.slice(0,6); numRed.each {|n| sumRedNum[n] += 1}
    }
    ((sumRedNum.sort_by{|k,v| v}.reverse).collect {|k,v| k}).slice(0,16)
end

def hitSSQ(no,hitNs)
    def hitDesc(redNum, blueNum)
        case [redNum, blueNum]
        when [6,1]; return "1st"
        when [6,0]; return "2nd"
        when [5,1]; return "3rd(3000)"
        when [5,0], [4,1]; return "4th(200)"
        when [4,0], [3,1]; return "5th(10)"
        when [2,1], [1,1], [0,1]; return "6th(5)"
        else return "X"
        end
    end

    File.open("ssqHitNum.txt", "a") do |fp|
        fp.write(no + " " + (hitNs.join " ") + "\n")
    end

    hitRed, hitBlue = hitNs.slice(0,6), hitNs.slice(6,1)
    print "No\tRed\tBlue\tResult\n"
    print "=============================================\n"
    File.foreach("ssqNum.txt") { |line|
        num = (line.split " ").map {|x| x.to_i}
        hitRedNum = num.slice(0,6).inject(0) {|s,x| (hitRed.include? x) ? s+=1 : s}
        hitBlueNum = num.slice(6,1).inject(0) {|s,x| (hitBlue.include? x) ? s+=1 : s}
        printf "%15s---> %2s %2s %s\n",line.chomp,hitRedNum,hitBlueNum,hitDesc(hitRedNum,hitBlueNum)
    }
end

#
#   stock
#
$SXF = 0.0015 #手续费
$YHS = 0.001  #印花费
$GHF = 1.0   #过户费
$RATIO = [0.236, 0.382, 0.5, 0.618, 0.809]

#计算股票盈利
def winG(qty,priceB,priceS)
	return qty * priceS * (1 -$SXF - $YHS) - 2 * $GHF - qty * priceB * (1 + $SXF)
end

#算权证盈利
def winQ(qty,priceB,priceS)
	return qty * priceS * (1 - $SXF) - 2 * $GHF - qty * priceB * (1 + $SXF)
end

def div618(priceLo, priceHi)
    ($RATIO.size-1).downto(0) {|i|
        price = priceLo + (priceHi - priceLo) * $RATIO[i]
        printf "-------%.3f   %.3f-------\n", $RATIO[i], price
    }
end

$NO_TRADE_DAYS = [
    Date.new(2010,1,1), Date.new(2010,1,2), Date.new(2010,1,3), #元旦
    Date.new(2010,2,13), Date.new(2010,2,14), Date.new(2010,2,15), Date.new(2010,2,16), Date.new(2010,2,17), #春节
    Date.new(2010,2,18), Date.new(2010,2,19), Date.new(2010,2,20), Date.new(2010,2,21),
    Date.new(2010,4,3), Date.new(2010,4,4), Date.new(2010,4,5), #清明
    Date.new(2010,5,1), Date.new(2010,5,2), Date.new(2010,5,3), #劳动
    Date.new(2010,6,12), Date.new(2010,6,13), Date.new(2010,6,14), Date.new(2010,6,15), Date.new(2010,6,16),  #端午
    Date.new(2010,9,22), Date.new(2010,9,23), Date.new(2010,9,24), Date.new(2010,9,25), Date.new(2010,9,26),  #中秋
    Date.new(2010,10,1), Date.new(2010,10,2), Date.new(2010,10,3), Date.new(2010,10,4), Date.new(2010,10,5),  #国庆
    Date.new(2010,10,6), Date.new(2010,10,7), Date.new(2010,10,9)
]

def is_trade_day(day1)
    if [6,0].include?(day1.wday) or $NO_TRADE_DAYS.include?(day1)
        return false
    else
        return true
    end
end

def str2date(dayStr)
    return Date.strptime(str=dayStr,fmt='%F')
end

def dateByNTday(dateStr, ntday)
    day1 = str2date(dateStr)
    count = 0
    while count < ntday
        day1 += 1
        count += 1 if is_trade_day(day1)
    end
    print day1
end

File: db/doc/haskell/Money.hs
{-# OPTIONS_GHC -cpp #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Money (
    winG, winQ, div618, stopLoss
   ,win_ssq, hit_ssq, his
   ,qachina, combLen, fibs, sieves
             )
where
import System.Random
import System.IO
--import System.Time
--import System.Directory
import System.Environment()
import System.Process
import System.Exit

import Text.Printf (printf)
--import Text.Regex

import Data.List
--import Data.IORef
--import Data.Time.Calendar
--import Data.Time.Calendar.WeekDate 

import qualified Control.Exception as Exception

import Control.Monad
import Control.Applicative

import Control.Arrow
import Control.Concurrent

import Foreign.Marshal.Alloc
import Foreign.Storable

#if __GLASGOW_HASKELL__ >= 610
catchAny :: forall a.
                  IO a -> (Exception.SomeException -> IO a) -> IO a
catchAny f h = Exception.catch f (\e -> h (e :: Exception.SomeException))
#else
catchAny = Exception.catch
#endif

-- 菲波纳契数列
--fibs = 0 : 1 : [ a + b | (a, b) <- zip fibs (tail fibs)]

fibs :: forall a. Num a => Int -> [a]
fibs n = take n $ fibgen 1 1

fibgen :: forall a. Num a => a -> a -> [a]
fibgen n1 n2 = n1 : fibgen n2 (n1+n2) 

--求解素数的一个无限数列方法：
sieves :: forall a. Integral a => Int -> [a]
sieves n = take n $ sieve []

sieve :: forall a. Integral a => [a] -> [a]
sieve [] = sieve [2..]
sieve (x:xs) = x : sieve (filter (\y -> y `rem` x /= 0) xs)

--排列组合 same as Data.List.subsequences
-- combination :: [a] -> [[a]]
-- combination [] =  [[]]
-- combination (x:xs) = concat [[(x:ys), ys] | ys <- combination xs]

-- permutation :: Eq a => [a] -> [[a]]
-- permutation [] = [[]]
-- permutation xs = concatMap (\x -> map (x:) $ permutation (delete x xs)) xs

combLen :: Int -> Int -> Int
combLen x len
  | x > len = product [x-len+1..x] `div` product [1..len]
  | otherwise = 0

qachina :: IO ()
qachina = do
    -- putStrLn "Running qachina..."
    -- runCommand "mighty mighty.conf mighty.route" >>= waitForProcess
    return ()
    --pid <- runCommand "mighty mighty.conf mighty.route"
    --waitForProcess pid >>= exitWith

------------------------------------------------------------------------------

sxf :: Float
sxf = 0.0015  --手续费

yhs :: Float
yhs = 0.001   --印花费

ghf :: Float
ghf = 1.0     --过户费

--计算股票盈利
winG :: Float -> Float -> Float -> Float
winG qty pb ps = qty * ps * (1 - sxf - yhs) - 2 * ghf - qty * pb * (1 + sxf)

--算权证盈利
winQ :: Float -> Float -> Float -> Float
winQ qty pb ps = qty * ps * (1 - sxf) - 2 * ghf - qty * pb * (1 + sxf)

--止损价
stopLoss :: Float -> Float -> Float -> IO ()
stopLoss qty pb lossRate = do
    let t = qty * pb * (1 + sxf)
    printf "Stop Loss at: %.2f\n" ((pb - (t * lossRate) / qty) :: Float)
    printf "Lost Money: %.2f\n"  ((t * lossRate) :: Float)

div618 :: Double -> Double -> IO ()
div618 p1 p2 = do
    let price r = if p1 <= p2 then p1 + (p2 - p1) * r else p1 - (p1 - p2) * r
        ratio = [0,0.191,0.236,0.382,0.5,0.618,0.809,1] 
    if p1 <= p2
        then  mapM (\r -> printf "-------%.3f   %.2f-------\n" r $ price r) $ reverse ratio
        else  mapM (\r -> printf "-------%.3f   %.2f-------\n" r $ price r) $ ratio
        
    return ()

-- fG y m d = fromGregorian y m d

-- no_trade_days = [fG 2011 1 1, fG 2011 1 2, fG 2011 1 3, --元旦
--                  fG 2011 2 2, fG 2011 2 3, fG 2011 2 4, fG 2011 2 5, fG 2011 2 6, --春节
--                  fG 2011 2 7, fG 2011 2 8,
--                  fG 2011 4 3, fG 2011 4 4, fG 2011 4 5, --清明
--                  fG 2011 4 30, fG 2011 5 1, fG 2011 5 2, --劳动
--                  fG 2011 6 4, fG 2011 6 5, fG 2011 6 6,  --端午
--                  fG 2013 9 19, fG 2013 9 20, fG 2013 9 21,  --中秋
--                  fG 2011 10 1, fG 2011 10 2, fG 2011 10 3, fG 2011 10 4, fG 2011 10 5,  --国庆
--                  fG 2011 10 6, fG 2011 10 7
--                 ]

-- is_trade_day day =
--     if (elem wd [6,7]) || (elem day no_trade_days) then False else True
--     where (y,w,wd) = toWeekDate day

--dateWindow sDayStr n = mapM_ (\nday -> dayByNTDay  sDayStr nday) $ fibs n
{--
dayByNTDay sDayStr nDay = do {
    let {
        [y,m,d] = map (\x -> read x::Int) $ splitRegex (mkRegex "-") sDayStr;
        sDay = fG (fromIntegral y) m d
    };
    print $ find_the_day sDay nDay ;
}
--}
-- find_the_day sDay 0 = sDay
-- find_the_day sDay nDay = 
--     let nextDay = addDays 1 sDay
--     in if (is_trade_day $ nextDay)
--     then find_the_day nextDay (nDay - 1)
--     else find_the_day nextDay nDay

------------------------------------------------------------------------
-- rollDice :: Int -> IO Int
-- rollDice n = do 
--     tmp <- doesFileExist "/dev/urandom" 
--     myGen <- if tmp
--         then mkStdGen <$> betterSeed
--         else (mkStdGen . fromInteger) <$> picoSec
    
--     return $ (take 1 $ randomRs (1,n) myGen) !! 0

-- picoSec :: IO Integer
-- picoSec = ctPicosec <$> (getClockTime >>= toCalendarTime)

betterSeed :: IO Int
betterSeed = alloca $ \p -> do
    withBinaryFile "/dev/urandom" ReadMode $ \h ->
        hGetBuf h p $ sizeOf (undefined :: Int)
    peek p

------------------------------------------------------------------------
win_ssq :: Int -> String -> String -> IO ()
win_ssq count noRed noBlue = do
    --let noStr = (map (\x->show x) noRed) ++ ["-"] ++ (map (\x->show x) noBlue)
    --writeFile "noRedBlue.txt" $ concat $ intersperse " " noStr
    let noRedLst =  map (\x -> read x::Int) $ words noRed
        noBlueLst = map (\x -> read x::Int) $ words noBlue

    --gRed <- goodRed

    _ <- setStdGen <$> (mkStdGen <$> betterSeed)
    okBlue <- pickNums ([1..16] \\ noBlueLst) count []
    result <- pickSSQ count
              ([1..33] \\ [1,33])
              ([1..33] \\ noRedLst)
              okBlue []
    --forM_ result (\x -> print x)
    writeFile ssqNum $ ints2str result

pickSSQ :: Int -> [Int] -> [Int] -> [Int] -> [[Int]] -> IO [[Int]]
pickSSQ 0 _ _ _ acc = return acc
pickSSQ 1 xRed _ okBlue acc = do
    red <- sort <$> pickNums xRed 6 []
    let one = red ++ [okBlue!!0]
    print one
    return $ reverse $ one : acc
pickSSQ count xRed yesRed okBlue acc = do
    red <- sort <$> pickNums yesRed 6 []
    let one = red ++ [okBlue!!(count-1)]
    print one
    pickSSQ (count-1) xRed yesRed okBlue $ one:acc

pick :: [a] -> IO a
pick xs = randomRIO (0, length xs - 1) >>= return . (xs !!)

pickNums :: [Int] -> Int -> [Int] -> IO [Int]
pickNums _ 0 acc = return acc
pickNums from count acc = do
  x <- pick from
  threadDelay 1000000 -- 1 second
  pickNums (from \\ [x]) (count-1) (x:acc)

ints2str :: [[Int]] -> String
ints2str ints = concat $ intersperse "\n" strLst
    where
        strLst = map (\x -> to_str x) ints
        to_str a = unwords $ map (\x -> show x) a 

-- goodRed:: IO [Int]
-- goodRed = do {
--     samp <- fmap (concat . str_ints_hit) $ readFile ssqHitNum;
--     return $ sort $ take 21 $ map (\(a,_) -> a) $ statis samp;
-- }

statis :: [Int] -> [(Int,Int)]
statis samp = map (\(a,b) -> (b,a)) $ times4n
    where
        times4n = (reverse . sort) $ map (length &&& head) $ (group . sort) samp

--skip the 期号
str_ints_hit :: String -> [[Int]]
str_ints_hit str = map (\line -> map (\x->read x::Int) line) $ map (init . tail . words) $ lines str

str_ints_pick :: String -> [[Int]]
str_ints_pick str = map (\line -> map (\x->read x::Int) line) $ map words $ lines str

hit_ssq :: String -> String -> IO ()
hit_ssq no hitNum = do
    let hitLst =  map (\x -> read x::Int) $ words hitNum
    let hitRed redNo = foldl (\acc x -> if(x `elem` (init hitLst)) then acc+1 else acc) 0 redNo
    dats <- readFile ssqHitNum
    if (any (\x -> no `elem` (words x)) (lines dats) == False)
        then catchAny (appendFile ssqHitNum $ unwords $ [no] ++ map (\n -> show n) hitLst ++ ["\n"]) (\e -> fail $ "Unable save hit numbers" ++ show e)
        else return ()

    --gRed <- goodRed
    --printf "good red hit %d\n" ((hitRed gRed) :: Int)

    nums <- fmap str_ints_pick $ readFile ssqNum
    forM_ nums (\n -> do
        let hitR = hitRed $ init n
            hitB = if (n!!6 == hitLst!!6) then 1 else 0 
        printf "%s\t%d:%d\t%s\n" (show n) (hitR::Int) (hitB::Int) (hit_desc hitR hitB))

hit_desc :: forall a a1.
                  (Eq a, Eq a1, Num a, Num a1) =>
                  a -> a1 -> [Char]
hit_desc red blue
    | red == 6 && blue == 1 = "First"
    | red == 6 && blue == 0 = "Second"
    | red == 5 && blue == 1 = "Third(3000)"
    | (red == 5 && blue == 0) || (red == 4 && blue == 1) = "Fourth(200)"
    | (red == 4 && blue == 0) || (red == 3 && blue == 1) = "Fifth(10)"
    | (blue == 1) = "Sixth(5)"
    | otherwise = "X" 

his :: IO ()
his = do
  pid <- runCommand $ "tail " ++ ssqHitNum
  waitForProcess pid >>= exitWith

ssqNum :: [Char]
ssqNum = "/media/D/qachina/db/doc/money/" ++ "ssqNum.txt"

ssqHitNum :: [Char]
ssqHitNum = "/media/D/qachina/db/doc/money/" ++ "ssqHitNum.txt"

------------------------------------------------------------------------
